#  PROJECT :INDIA GENERAL ELECTIONS RESULT ANALYSIS 2024


# Executed this project using SQL within Jupyter Notebook, integrated with Python.



# IMPORTING THE PANDAS LIBRARIES 

In [5]:
import pandas as pd
import numpy as np 


# CONNECTING WITH SQL SERVER 

In [7]:
import mysql.connector

In [8]:
!pip install mysql-connector-python

Defaulting to user installation because normal site-packages is not writeable


In [10]:
from sqlalchemy import create_engine

In [11]:
mydb = mysql.connector.connect(
    host='localhost',
    user='root',
    password='password',
    database='india_general_elections_2024',  # fix typo here
    use_pure=True)

In [12]:
mydb

In [13]:
pd.read_sql_query("""SELECT * FROM india_general_elections_2024.constituencywise_details;""",mydb)

C:\Users\Mankirat\AppData\Local\Temp\ipykernel_22588\3587043906.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query("""SELECT * FROM india_general_elections_2024.constituencywise_details;""",mydb)


S.N.                      Candidate                     Party  EVM Votes  \
0       1              DR JITENDRA SINGH    Bharatiya Janata Party     566833   
1       2                   CH LAL SINGH  Indian National Congress     444151   
2       3            GHULAM MOHD SAROORI               Independent      39519   
3       4                     MEHRAJ DIN               Independent       9055   
4       5                     AMIT KUMAR       Bahujan Samaj Party       8559   
..    ...                            ...                       ...        ...   
507    27  SALIMUDDIN KHALILUDDIN SHAIKH               Independent        370   
508    28      DR. SACHIN SAHEBRAO PATIL               Independent        346   
509    29                           NOTA         None of the Above      11603   
510     1               BHAJAN LAL JATAV  Indian National Congress     525379   
511     2                      INDU DEVI    Bharatiya Janata Party     426798   

     Postal Votes  Total Votes  % of Votes Constituency ID  
0            4243       571076       51.28            U084  
1            2552       446703       40.11            U084  
2              80        39599        3.56            U084  
3              27         9082        0.82            U084  
4              83         8642        0.78            U084  
..            ...          ...         ...             ...  
507             1          371        0.04           S1324  
508             5          351        0.03           S1324  
509            83        11686        1.12           S1324  
510          4632       530011       53.64           S2010  
511          4268       431066       43.62           S2010  

[512 rows x 8 columns]

In [25]:
engine = create_engine("mysql+pymysql://root:password@localhost/india_general_elections_2024")


In [27]:
pd.read_sql_query("""SELECT * FROM india_general_elections_2024.constituencywise_details;""",engine)

S.N.                      Candidate                     Party  EVM Votes  \
0       1              DR JITENDRA SINGH    Bharatiya Janata Party     566833   
1       2                   CH LAL SINGH  Indian National Congress     444151   
2       3            GHULAM MOHD SAROORI               Independent      39519   
3       4                     MEHRAJ DIN               Independent       9055   
4       5                     AMIT KUMAR       Bahujan Samaj Party       8559   
..    ...                            ...                       ...        ...   
507    27  SALIMUDDIN KHALILUDDIN SHAIKH               Independent        370   
508    28      DR. SACHIN SAHEBRAO PATIL               Independent        346   
509    29                           NOTA         None of the Above      11603   
510     1               BHAJAN LAL JATAV  Indian National Congress     525379   
511     2                      INDU DEVI    Bharatiya Janata Party     426798   

     Postal Votes  Total Votes  % of Votes Constituency ID  
0            4243       571076       51.28            U084  
1            2552       446703       40.11            U084  
2              80        39599        3.56            U084  
3              27         9082        0.82            U084  
4              83         8642        0.78            U084  
..            ...          ...         ...             ...  
507             1          371        0.04           S1324  
508             5          351        0.03           S1324  
509            83        11686        1.12           S1324  
510          4632       530011       53.64           S2010  
511          4268       431066       43.62           S2010  

[512 rows x 8 columns]


# Task1:  Total number of seats in Parliament Constituency

In [272]:
df1=pd.read_sql_query(""" SELECT DISTINCT
    COUNT(`Parliament Constituency`)
FROM
    constituencywise_results;""",engine)


In [274]:
df1


COUNT(`Parliament Constituency`)
0                               543

# conclusion :Parliament Constituency have total 543 Seats

  # -- Task 2 : what are the total number of seats available for elections for each states 

In [282]:
T2=pd.read_sql_query(""" select s.State as State_name , count(cr.`Constituency ID`) as total__seats_available
    from states as s 
    join state_results as sr
    on s.`State ID`=sr.`State ID` 
    join constituencywise_results as cr
    on cr.`Parliament Constituency` = sr.`Parliament Constituency` 
    group by State_name 
    order by State_name ;""" ,engine)

In [284]:
T2

State_name  total__seats_available
0              Andaman & Nicobar Islands                       1
1                         Andhra Pradesh                      25
2                      Arunachal Pradesh                       2
3                                  Assam                      14
4                                  Bihar                      40
5                             Chandigarh                       1
6                           Chhattisgarh                      11
7   Dadra & Nagar Haveli and Daman & Diu                       2
8                                  Delhi                       7
9                                    Goa                       2
10                               Gujarat                      26
11                               Haryana                      10
12                      Himachal Pradesh                       4
13                     Jammu and Kashmir                       5
14                             Jharkhand                      14
15                             Karnataka                      28
16                                Kerala                      20
17                                Ladakh                       1
18                           Lakshadweep                       1
19                        Madhya Pradesh                      29
20                           Maharashtra                      48
21                               Manipur                       2
22                             Meghalaya                       2
23                               Mizoram                       1
24                              Nagaland                       1
25                                Odisha                      21
26                            Puducherry                       1
27                                Punjab                      13
28                             Rajasthan                      25
29                                Sikkim                       1
30                            Tamil Nadu                      39
31                             Telangana                      17
32                               Tripura                       2
33                         Uttar Pradesh                      80
34                           Uttarakhand                       5
35                           West Bengal                      42

 # -- Task 3 : Total seats won by NDA alliance 

In [286]:


T3=pd.read_sql_query("""SELECT 
    SUM(CASE
        WHEN
            party IN ('Bharatiya Janata Party - BJP' , 'Telugu Desam - TDP',
                'Janata Dal  (United) - JD(U)',
                'Shiv Sena - SHS',
                'AJSU Party - AJSUP',
                'Apna Dal (Soneylal) - ADAL',
                'Asom Gana Parishad - AGP',
                'Hindustani Awam Morcha (Secular) - HAMS',
                'Janasena Party - JnP',
                'Janata Dal  (Secular) - JD(S)',
                'Lok Janshakti Party(Ram Vilas) - LJPRV',
                'Nationalist Congress Party - NCP',
                'Rashtriya Lok Dal - RLD',
                'Sikkim Krantikari Morcha - SKM')
        THEN
            `Won`
        ELSE 0
    END) AS NDA_Total_Seats_Won
FROM
    partywise_results;
""",engine)


In [288]:
T3

NDA_Total_Seats_Won
0                292.0

# CONCLUSION  :Total seats won by NDA alliance was 292 seats 

 # --  Task 4 :  Total seats won by NDA alliance parties

In [292]:
T4=pd.read_sql_query("""  SELECT 
    party as Party_Name,
    won as Seats_Won
FROM 
    partywise_results
WHERE 
    party IN (
        'Bharatiya Janata Party - BJP', 
        'Telugu Desam - TDP', 
		'Janata Dal  (United) - JD(U)',
        'Shiv Sena - SHS', 
        'AJSU Party - AJSUP', 
        'Apna Dal (Soneylal) - ADAL', 
        'Asom Gana Parishad - AGP',
        'Hindustani Awam Morcha (Secular) - HAMS', 
        'Janasena Party - JnP', 
		'Janata Dal  (Secular) - JD(S)',
        'Lok Janshakti Party(Ram Vilas) - LJPRV', 
        'Nationalist Congress Party - NCP',
        'Rashtriya Lok Dal - RLD', 
        'Sikkim Krantikari Morcha - SKM'
    )
ORDER BY Seats_Won DESC;""",engine)


In [294]:
T4

Party_Name  Seats_Won
0              Bharatiya Janata Party - BJP        240
1                        Telugu Desam - TDP         16
2              Janata Dal  (United) - JD(U)         12
3                           Shiv Sena - SHS          7
4    Lok Janshakti Party(Ram Vilas) - LJPRV          5
5             Janata Dal  (Secular) - JD(S)          2
6                      Janasena Party - JnP          2
7                   Rashtriya Lok Dal - RLD          2
8                  Asom Gana Parishad - AGP          1
9                        AJSU Party - AJSUP          1
10  Hindustani Awam Morcha (Secular) - HAMS          1
11         Nationalist Congress Party - NCP          1
12           Sikkim Krantikari Morcha - SKM          1
13               Apna Dal (Soneylal) - ADAL          1

# CONCLUSION : NDA alliance parties like BJP won by 240 


# -- Task 5 : Total Seats Won by I.N.D.I.A. Allianz 

In [299]:
T5=pd.read_sql_query("""
SELECT 
    SUM(CASE 
            WHEN party IN (
                'Indian National Congress - INC',
                'Aam Aadmi Party - AAAP',
                'All India Trinamool Congress - AITC',
                'Bharat Adivasi Party - BHRTADVSIP',
                'Communist Party of India  (Marxist) - CPI(M)',
                'Communist Party of India  (Marxist-Leninist)  (Liberation) - CPI(ML)(L)',
                'Communist Party of India - CPI',
                'Dravida Munnetra Kazhagam - DMK',
                'Indian Union Muslim League - IUML',
                'Nat`Jammu & Kashmir National Conference - JKN',
                'Jharkhand Mukti Morcha - JMM',
                'Jammu & Kashmir National Conference - JKN',
                'Kerala Congress - KEC',
                'Marumalarchi Dravida Munnetra Kazhagam - MDMK',
                'Nationalist Congress Party Sharadchandra Pawar - NCPSP',
                'Rashtriya Janata Dal - RJD',
                'Rashtriya Loktantrik Party - RLTP',
                'Revolutionary Socialist Party - RSP',
                'Samajwadi Party - SP',
                'Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT',
                'Viduthalai Chiruthaigal Katchi - VCK'
            ) THEN `Won`
            ELSE 0 
        END) AS INDIA_Total_Seats_Won
FROM 
    partywise_results ;""",engine)

In [301]:
T5

INDIA_Total_Seats_Won
0                  234.0

# -- Task 6 : Seats Won by I.N.D.I.A. Allianz Parties

In [304]:
T6=pd.read_sql_query("""SELECT 
    party as Party_Name,
    won as Seats_Won
FROM 
    partywise_results
WHERE 
    party IN (
        'Indian National Congress - INC',
                'Aam Aadmi Party - AAAP',
                'All India Trinamool Congress - AITC',
                'Bharat Adivasi Party - BHRTADVSIP',
                'Communist Party of India  (Marxist) - CPI(M)',
                'Communist Party of India  (Marxist-Leninist)  (Liberation) - CPI(ML)(L)',
                'Communist Party of India - CPI',
                'Dravida Munnetra Kazhagam - DMK',
                'Indian Union Muslim League - IUML',
                'Nat`Jammu & Kashmir National Conference - JKN',
                'Jharkhand Mukti Morcha - JMM',
                'Jammu & Kashmir National Conference - JKN',
                'Kerala Congress - KEC',
                'Marumalarchi Dravida Munnetra Kazhagam - MDMK',
                'Nationalist Congress Party Sharadchandra Pawar - NCPSP',
                'Rashtriya Janata Dal - RJD',
                'Rashtriya Loktantrik Party - RLTP',
                'Revolutionary Socialist Party - RSP',
                'Samajwadi Party - SP',
                'Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT',
                'Viduthalai Chiruthaigal Katchi - VCK'
    )
ORDER BY Seats_Won DESC ;
""",engine)


In [306]:
T6

Party_Name  Seats_Won
0                      Indian National Congress - INC         99
1                                Samajwadi Party - SP         37
2                 All India Trinamool Congress - AITC         29
3                     Dravida Munnetra Kazhagam - DMK         22
4      Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT          9
5   Nationalist Congress Party Sharadchandra Pawar...          8
6        Communist Party of India  (Marxist) - CPI(M)          4
7                          Rashtriya Janata Dal - RJD          4
8                              Aam Aadmi Party - AAAP          3
9                   Indian Union Muslim League - IUML          3
10                       Jharkhand Mukti Morcha - JMM          3
11                     Communist Party of India - CPI          2
12  Communist Party of India  (Marxist-Leninist)  ...          2
13          Jammu & Kashmir National Conference - JKN          2
14               Viduthalai Chiruthaigal Katchi - VCK          2
15                              Kerala Congress - KEC          1
16      Marumalarchi Dravida Munnetra Kazhagam - MDMK          1
17                Revolutionary Socialist Party - RSP          1
18                  Rashtriya Loktantrik Party - RLTP          1
19                  Bharat Adivasi Party - BHRTADVSIP          1

# TASK 7  Add new column field in table partywise_results to get the Party Allianz as NDA, I.N.D.I.A and OTHER

#  FIRST IMPORT LIBRARY "sqlalchemy" , so that we can add new column into the given data set  

In [106]:
from sqlalchemy import tex


# firstly alter the table name then add the column according to the task 

In [108]:
with engine.connect() as conn:
    conn.execute(text("""
        ALTER TABLE partywise_results
        ADD party_alliance VARCHAR(50);
    """))


In [115]:
df=pd.read_sql_query("""SELECT * FROM partywise_results ;""",engine)

In [117]:
df

Party  Won  Party ID  \
0                              Aam Aadmi Party - AAAP    3         1   
1                            Asom Gana Parishad - AGP    1        83   
2      All India Majlis-E-Ittehadul Muslimeen - AIMIM    1       118   
3                 All India Trinamool Congress - AITC   29       140   
4                                  AJSU Party - AJSUP    1       160   
5                        Bharatiya Janata Party - BJP  240       369   
6                      Communist Party of India - CPI    2       544   
7   Communist Party of India  (Marxist-Leninist)  ...    2       545   
8        Communist Party of India  (Marxist) - CPI(M)    4       547   
9                     Dravida Munnetra Kazhagam - DMK   22       582   
10            Hindustani Awam Morcha (Secular) - HAMS    1       664   
11                     Indian National Congress - INC   99       742   
12                                  Independent - IND    7       743   
13                  Indian Union Muslim League - IUML    3       772   
14                      Janata Dal  (Secular) - JD(S)    2       804   
15                       Janata Dal  (United) - JD(U)   12       805   
16          Jammu & Kashmir National Conference - JKN    2       834   
17                       Jharkhand Mukti Morcha - JMM    3       852   
18                               Janasena Party - JnP    2       860   
19                              Kerala Congress - KEC    1       911   
20      Marumalarchi Dravida Munnetra Kazhagam - MDMK    1      1046   
21                   Nationalist Congress Party - NCP    1      1142   
22                         Rashtriya Janata Dal - RJD    4      1420   
23                            Rashtriya Lok Dal - RLD    2      1458   
24                Revolutionary Socialist Party - RSP    1      1534   
25                          Shiromani Akali Dal - SAD    1      1584   
26                     Sikkim Krantikari Morcha - SKM    1      1658   
27                               Samajwadi Party - SP   37      1680   
28                                 Telugu Desam - TDP   16      1745   
29               Viduthalai Chiruthaigal Katchi - VCK    2      1847   
30      Yuvajana Sramika Rythu Congress Party - YSRCP    4      1888   
31            United People’s Party, Liberal - UPPL    1      1998   
32                         Apna Dal (Soneylal) - ADAL    1      2070   
33                  Rashtriya Loktantrik Party - RLTP    1      2484   
34                    Zoram People’s Movement - ZPM    1      2757   
35             Aazad Samaj Party (Kanshi Ram) - ASPKR    1      2989   
36             Lok Janshakti Party(Ram Vilas) - LJPRV    5      3165   
37     Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT    9      3369   
38                  Voice of the People Party - VOTPP    1      3388   
39                  Bharat Adivasi Party - BHRTADVSIP    1      3482   
40                                    Shiv Sena - SHS    7      3529   
41  Nationalist Congress Party Sharadchandra Pawar...    8      3620   

   party_alliance  
0            None  
1            None  
2            None  
3            None  
4            None  
5            None  
6            None  
7            None  
8            None  
9            None  
10           None  
11           None  
12           None  
13           None  
14           None  
15           None  
16           None  
17           None  
18           None  
19           None  
20           None  
21           None  
22           None  
23           None  
24           None  
25           None  
26           None  
27           None  
28           None  
29           None  
30           None  
31           None  
32           None  
33           None  
34           None  
35           None  
36           None  
37           None  
38           None  
39           None  
40           None  
41           None

#  new column name "party-alliance" is created and now insert tha data into it 


# add th data of I.N.D.I.A Allianz

In [134]:


with engine.connect() as conn:
    conn.execute(text("""
        UPDATE partywise_results
        SET party_alliance = 'I.N.D.I.A'
        WHERE party IN (
            'Indian National Congress - INC',
            'Aam Aadmi Party - AAAP',
            'All India Trinamool Congress - AITC',
            'Bharat Adivasi Party - BHRTADVSIP',
            'Communist Party of India  (Marxist) - CPI(M)',
            'Communist Party of India  (Marxist-Leninist)  (Liberation) - CPI(ML)(L)',
            'Communist Party of India - CPI',
            'Dravida Munnetra Kazhagam - DMK',	
            'Indian Union Muslim League - IUML',
            'Jammu & Kashmir National Conference - JKN',
            'Jharkhand Mukti Morcha - JMM',
            'Kerala Congress - KEC',
            'Marumalarchi Dravida Munnetra Kazhagam - MDMK',
            'Nationalist Congress Party Sharadchandra Pawar - NCPSP',
            'Rashtriya Janata Dal - RJD',
            'Rashtriya Loktantrik Party - RLTP',
            'Revolutionary Socialist Party - RSP',
            'Samajwadi Party - SP',
            'Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT',
            'Viduthalai Chiruthaigal Katchi - VCK'
        );
    """))


In [154]:
df1=pd.read_sql_query("""SELECT * FROM partywise_results WHERE party_alliance = 'I.N.D.I.A' ;""",engine)

In [156]:
df1



Party  Won  Party ID  \
0                              Aam Aadmi Party - AAAP    3         1   
1                 All India Trinamool Congress - AITC   29       140   
2                      Communist Party of India - CPI    2       544   
3   Communist Party of India  (Marxist-Leninist)  ...    2       545   
4        Communist Party of India  (Marxist) - CPI(M)    4       547   
5                     Dravida Munnetra Kazhagam - DMK   22       582   
6                      Indian National Congress - INC   99       742   
7                   Indian Union Muslim League - IUML    3       772   
8           Jammu & Kashmir National Conference - JKN    2       834   
9                        Jharkhand Mukti Morcha - JMM    3       852   
10                              Kerala Congress - KEC    1       911   
11      Marumalarchi Dravida Munnetra Kazhagam - MDMK    1      1046   
12                         Rashtriya Janata Dal - RJD    4      1420   
13                Revolutionary Socialist Party - RSP    1      1534   
14                               Samajwadi Party - SP   37      1680   
15               Viduthalai Chiruthaigal Katchi - VCK    2      1847   
16                  Rashtriya Loktantrik Party - RLTP    1      2484   
17     Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT    9      3369   
18                  Bharat Adivasi Party - BHRTADVSIP    1      3482   
19  Nationalist Congress Party Sharadchandra Pawar...    8      3620   

   party_alliance  
0       I.N.D.I.A  
1       I.N.D.I.A  
2       I.N.D.I.A  
3       I.N.D.I.A  
4       I.N.D.I.A  
5       I.N.D.I.A  
6       I.N.D.I.A  
7       I.N.D.I.A  
8       I.N.D.I.A  
9       I.N.D.I.A  
10      I.N.D.I.A  
11      I.N.D.I.A  
12      I.N.D.I.A  
13      I.N.D.I.A  
14      I.N.D.I.A  
15      I.N.D.I.A  
16      I.N.D.I.A  
17      I.N.D.I.A  
18      I.N.D.I.A  
19      I.N.D.I.A

In [144]:
df = pd.read_sql_query("SELECT DISTINCT party FROM partywise_results;", engine)
print(df)


                                                party
0                              Aam Aadmi Party - AAAP
1                            Asom Gana Parishad - AGP
2      All India Majlis-E-Ittehadul Muslimeen - AIMIM
3                 All India Trinamool Congress - AITC
4                                  AJSU Party - AJSUP
5                        Bharatiya Janata Party - BJP
6                      Communist Party of India - CPI
7   Communist Party of India  (Marxist-Leninist)  ...
8        Communist Party of India  (Marxist) - CPI(M)
9                     Dravida Munnetra Kazhagam - DMK
10            Hindustani Awam Morcha (Secular) - HAMS
11                     Indian National Congress - INC
12                                  Independent - IND
13                  Indian Union Muslim League - IUML
14                      Janata Dal  (Secular) - JD(S)
15                       Janata Dal  (United) - JD(U)
16          Jammu & Kashmir National Conference - JKN
17                       Jha

In [150]:
df = pd.read_sql_query("SELECT party, party_alliance FROM partywise_results WHERE party_alliance = 'I.N.D.I.A';", engine)
df


party party_alliance
0                              Aam Aadmi Party - AAAP      I.N.D.I.A
1                 All India Trinamool Congress - AITC      I.N.D.I.A
2                      Communist Party of India - CPI      I.N.D.I.A
3   Communist Party of India  (Marxist-Leninist)  ...      I.N.D.I.A
4        Communist Party of India  (Marxist) - CPI(M)      I.N.D.I.A
5                     Dravida Munnetra Kazhagam - DMK      I.N.D.I.A
6                      Indian National Congress - INC      I.N.D.I.A
7                   Indian Union Muslim League - IUML      I.N.D.I.A
8           Jammu & Kashmir National Conference - JKN      I.N.D.I.A
9                        Jharkhand Mukti Morcha - JMM      I.N.D.I.A
10                              Kerala Congress - KEC      I.N.D.I.A
11      Marumalarchi Dravida Munnetra Kazhagam - MDMK      I.N.D.I.A
12                         Rashtriya Janata Dal - RJD      I.N.D.I.A
13                Revolutionary Socialist Party - RSP      I.N.D.I.A
14                               Samajwadi Party - SP      I.N.D.I.A
15               Viduthalai Chiruthaigal Katchi - VCK      I.N.D.I.A
16                  Rashtriya Loktantrik Party - RLTP      I.N.D.I.A
17     Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT      I.N.D.I.A
18                  Bharat Adivasi Party - BHRTADVSIP      I.N.D.I.A
19  Nationalist Congress Party Sharadchandra Pawar...      I.N.D.I.A

# --  adding the dada of NDA Allianz into new column which is created in above


In [176]:
from sqlalchemy import text

with engine.begin() as conn:  # ensures commit
    conn.execute(text("""
        UPDATE partywise_results
        SET party_alliance = 'NDA'
        WHERE party IN (
            'Bharatiya Janata Party - BJP',
            'Telugu Desam - TDP',
            'Janata Dal  (United) - JD(U)',
            'Shiv Sena - SHS',
            'AJSU Party - AJSUP',
            'Apna Dal (Soneylal) - ADAL',
            'Asom Gana Parishad - AGP',
            'Hindustani Awam Morcha (Secular) - HAMS',
            'Janasena Party - JnP',
            'Janata Dal  (Secular) - JD(S)',
            'Lok Janshakti Party(Ram Vilas) - LJPRV',
            'Nationalist Congress Party - NCP',
            'Rashtriya Lok Dal - RLD',
            'Sikkim Krantikari Morcha - SKM'
        );
    """))


In [182]:
df = pd.read_sql_query("""
    SELECT party, party_alliance 
    FROM partywise_results 
    WHERE party_alliance = 'NDA';
""", engine)

df


party party_alliance
0                  Asom Gana Parishad - AGP            NDA
1                        AJSU Party - AJSUP            NDA
2              Bharatiya Janata Party - BJP            NDA
3   Hindustani Awam Morcha (Secular) - HAMS            NDA
4             Janata Dal  (Secular) - JD(S)            NDA
5              Janata Dal  (United) - JD(U)            NDA
6                      Janasena Party - JnP            NDA
7          Nationalist Congress Party - NCP            NDA
8                   Rashtriya Lok Dal - RLD            NDA
9            Sikkim Krantikari Morcha - SKM            NDA
10                       Telugu Desam - TDP            NDA
11               Apna Dal (Soneylal) - ADAL            NDA
12   Lok Janshakti Party(Ram Vilas) - LJPRV            NDA
13                          Shiv Sena - SHS            NDA

#  inserting tha data of OTHERS party into the column

In [188]:


with engine.begin() as conn:  # ensures the update is committed
    conn.execute(text("""
        UPDATE partywise_results
        SET party_alliance = 'OTHER'
        WHERE party_alliance IS NULL;
    """))


In [190]:
df = pd.read_sql_query("""
    SELECT party, party_alliance
    FROM partywise_results
    WHERE party_alliance = 'OTHER';
""", engine)

df.head()


party party_alliance
0  All India Majlis-E-Ittehadul Muslimeen - AIMIM          OTHER
1                               Independent - IND          OTHER
2                       Shiromani Akali Dal - SAD          OTHER
3   Yuvajana Sramika Rythu Congress Party - YSRCP          OTHER
4         United People’s Party, Liberal - UPPL          OTHER

In [200]:
df = pd.read_sql_query("""SELECT * FROM partywise_results ;""", engine)


In [202]:
df

Party  Won  Party ID  \
0                              Aam Aadmi Party - AAAP    3         1   
1                            Asom Gana Parishad - AGP    1        83   
2      All India Majlis-E-Ittehadul Muslimeen - AIMIM    1       118   
3                 All India Trinamool Congress - AITC   29       140   
4                                  AJSU Party - AJSUP    1       160   
5                        Bharatiya Janata Party - BJP  240       369   
6                      Communist Party of India - CPI    2       544   
7   Communist Party of India  (Marxist-Leninist)  ...    2       545   
8        Communist Party of India  (Marxist) - CPI(M)    4       547   
9                     Dravida Munnetra Kazhagam - DMK   22       582   
10            Hindustani Awam Morcha (Secular) - HAMS    1       664   
11                     Indian National Congress - INC   99       742   
12                                  Independent - IND    7       743   
13                  Indian Union Muslim League - IUML    3       772   
14                      Janata Dal  (Secular) - JD(S)    2       804   
15                       Janata Dal  (United) - JD(U)   12       805   
16          Jammu & Kashmir National Conference - JKN    2       834   
17                       Jharkhand Mukti Morcha - JMM    3       852   
18                               Janasena Party - JnP    2       860   
19                              Kerala Congress - KEC    1       911   
20      Marumalarchi Dravida Munnetra Kazhagam - MDMK    1      1046   
21                   Nationalist Congress Party - NCP    1      1142   
22                         Rashtriya Janata Dal - RJD    4      1420   
23                            Rashtriya Lok Dal - RLD    2      1458   
24                Revolutionary Socialist Party - RSP    1      1534   
25                          Shiromani Akali Dal - SAD    1      1584   
26                     Sikkim Krantikari Morcha - SKM    1      1658   
27                               Samajwadi Party - SP   37      1680   
28                                 Telugu Desam - TDP   16      1745   
29               Viduthalai Chiruthaigal Katchi - VCK    2      1847   
30      Yuvajana Sramika Rythu Congress Party - YSRCP    4      1888   
31            United People’s Party, Liberal - UPPL    1      1998   
32                         Apna Dal (Soneylal) - ADAL    1      2070   
33                  Rashtriya Loktantrik Party - RLTP    1      2484   
34                    Zoram People’s Movement - ZPM    1      2757   
35             Aazad Samaj Party (Kanshi Ram) - ASPKR    1      2989   
36             Lok Janshakti Party(Ram Vilas) - LJPRV    5      3165   
37     Shiv Sena (Uddhav Balasaheb Thackrey) - SHSUBT    9      3369   
38                  Voice of the People Party - VOTPP    1      3388   
39                  Bharat Adivasi Party - BHRTADVSIP    1      3482   
40                                    Shiv Sena - SHS    7      3529   
41  Nationalist Congress Party Sharadchandra Pawar...    8      3620   

   party_alliance  
0       I.N.D.I.A  
1             NDA  
2           OTHER  
3       I.N.D.I.A  
4             NDA  
5             NDA  
6       I.N.D.I.A  
7       I.N.D.I.A  
8       I.N.D.I.A  
9       I.N.D.I.A  
10            NDA  
11      I.N.D.I.A  
12          OTHER  
13      I.N.D.I.A  
14            NDA  
15            NDA  
16      I.N.D.I.A  
17      I.N.D.I.A  
18            NDA  
19      I.N.D.I.A  
20      I.N.D.I.A  
21            NDA  
22      I.N.D.I.A  
23            NDA  
24      I.N.D.I.A  
25          OTHER  
26            NDA  
27      I.N.D.I.A  
28            NDA  
29      I.N.D.I.A  
30          OTHER  
31          OTHER  
32            NDA  
33      I.N.D.I.A  
34          OTHER  
35          OTHER  
36            NDA  
37      I.N.D.I.A  
38          OTHER  
39      I.N.D.I.A  
40            NDA  
41      I.N.D.I.A

# Task 8 :Which party alliance (NDA, I.N.D.I.A, or OTHER) won the most seats across all states?


In [319]:
T8 = pd.read_sql_query("""SELECT 
    p.party_alliance,
    COUNT(cr.`Constituency ID`) AS Seats_Won
FROM 
    constituencywise_results cr
JOIN 
    partywise_results p 
    ON cr.`Party ID` = p.`Party ID`
GROUP BY 
    p.party_alliance
ORDER BY 
    Seats_Won DESC;
""", engine)

In [321]:
T8

party_alliance  Seats_Won
0            NDA        292
1      I.N.D.I.A        234
2          OTHER         17

#  Task 9 : Winning candidate's name, their party name, total votes, and the margin of victory for a specific state and constituency?


In [324]:
T9 = pd.read_sql_query("""SELECT cr.`Winning Candidate`,
p.`Party`,
p.`party_alliance`, 
cr.`Total Votes`, 
cr.`Margin`, 
cr.`Constituency Name`,
s.State
FROM constituencywise_results as cr
JOIN partywise_results as p
ON cr.`Party ID` = p.`Party ID`
JOIN state_results as sr
ON cr.`Parliament Constituency` = sr.`Parliament Constituency`
JOIN states as s 
ON sr.`State ID` = s.`State ID`
WHERE s.State = 'Uttar Pradesh'
AND 
cr.`Constituency Name` = 'AMETHI';""", engine)

In [326]:
T9

Winning Candidate                           Party party_alliance  \
0       KISHORI LAL  Indian National Congress - INC      I.N.D.I.A   

   Total Votes  Margin Constituency Name          State  
0       539228  167196            AMETHI  Uttar Pradesh

#  Task 10 : What is the distribution of EVM votes versus postal votes for candidates in a specific constituency?

In [334]:
T10 = pd.read_sql_query(""" 
SELECT 
    cd.`EVM Votes`,
    cd.`Postal Votes`,
    cd.`Candidate`,
    cd.`Total Votes`,
    cr.`Constituency Name`
FROM
    constituencywise_details AS cd
        JOIN
    constituencywise_results AS cr ON cr.`Constituency ID` = cd.`Constituency ID`
WHERE
    cr.`Constituency Name` = 'ETAH'
ORDER BY cd.`Total Votes` DESC;


""", engine)

In [336]:
T10


EVM Votes  Postal Votes                    Candidate  Total Votes  \
0      474263          1545                DEVESH SHAKYA       475808   
1      446748          1008  RAJVEER SINGH (RAJU BHAIYA)       447756   
2       71407           178      MOHAMMAD IRFAN ADVOCATE        71585   
3        5145            11                         NOTA         5156   
4        3944             5                 ANUPAM KUMAR         3949   
5        1728             0                KAILASH KUMAR         1728   
6        1261             4                      DANVEER         1265   
7         985             7              UDAY VEER SINGH          992   
8         777             3                  ASHOK KUMAR          780   
9         742             5               VIABHAV MISHRA          747   
10        609             6                   AMAR SINGH          615   

   Constituency Name  
0               ETAH  
1               ETAH  
2               ETAH  
3               ETAH  
4               ETAH  
5               ETAH  
6               ETAH  
7               ETAH  
8               ETAH  
9               ETAH  
10              ETAH


#  Task 11 : Which parties won the most seats in s State, and how many seats did each party win?


In [339]:
T11= pd.read_sql_query("""select pr.`party`,
count(cr.`Constituency ID`) as Seats_Won
from partywise_results as pr
join constituencywise_results aS cr
on cr.`Party ID`=pr.`Party ID`
join state_results as sr
on sr.`Parliament Constituency` = cr.`Parliament Constituency`
join states as s
on s.`State ID`=sr.`State ID`
where s.State="Andhra Pradesh"
group by 1
order by 2 desc;
""", engine)

In [341]:
T11

party  Seats_Won
0                             Telugu Desam - TDP         16
1  Yuvajana Sramika Rythu Congress Party - YSRCP          4
2                   Bharatiya Janata Party - BJP          3
3                           Janasena Party - JnP          2

# Task  12 : at is the total number of seats won by each party alliance (NDA, I.N.D.I.A, and OTHER) in each state for the India Elections 2024
 

In [344]:
T12 = pd.read_sql_query("""select s.State as state_name ,
sum(
case 
when pr.party_alliance="I.N.D.I.A" THEN 1 ELSE 0 END ) AS INDIA_Seats_Won,
Sum(
case 
when pr.party_alliance="NDA" THEN 1 ELSE 0 END ) AS NDA_Seats_Won,
sum(
case 
when pr.party_alliance="OTHER" THEN 1 ELSE 0 END ) AS OTHER_Seats_Won
From partywise_results as pr
join constituencywise_results aS cr
on cr.`Party ID`=pr.`Party ID`
join state_results as sr
on sr.`Parliament Constituency` = cr.`Parliament Constituency`
join states as s
on s.`State ID`=sr.`State ID`
group by 1
order by 1;
""", engine)

In [346]:
T12

state_name  INDIA_Seats_Won  NDA_Seats_Won  \
0              Andaman & Nicobar Islands              0.0            1.0   
1                         Andhra Pradesh              0.0           21.0   
2                      Arunachal Pradesh              0.0            2.0   
3                                  Assam              3.0           10.0   
4                                  Bihar              9.0           30.0   
5                             Chandigarh              1.0            0.0   
6                           Chhattisgarh              1.0           10.0   
7   Dadra & Nagar Haveli and Daman & Diu              0.0            1.0   
8                                  Delhi              0.0            7.0   
9                                    Goa              1.0            1.0   
10                               Gujarat              1.0           25.0   
11                               Haryana              5.0            5.0   
12                      Himachal Pradesh              0.0            4.0   
13                     Jammu and Kashmir              2.0            2.0   
14                             Jharkhand              5.0            9.0   
15                             Karnataka              9.0           19.0   
16                                Kerala             19.0            1.0   
17                                Ladakh              0.0            0.0   
18                           Lakshadweep              1.0            0.0   
19                        Madhya Pradesh              0.0           29.0   
20                           Maharashtra             30.0           17.0   
21                               Manipur              2.0            0.0   
22                             Meghalaya              1.0            0.0   
23                               Mizoram              0.0            0.0   
24                              Nagaland              1.0            0.0   
25                                Odisha              1.0           20.0   
26                            Puducherry              1.0            0.0   
27                                Punjab             10.0            0.0   
28                             Rajasthan             11.0           14.0   
29                                Sikkim              0.0            1.0   
30                            Tamil Nadu             39.0            0.0   
31                             Telangana              8.0            8.0   
32                               Tripura              0.0            2.0   
33                         Uttar Pradesh             43.0           36.0   
34                           Uttarakhand              0.0            5.0   
35                           West Bengal             30.0           12.0   

    OTHER_Seats_Won  
0               0.0  
1               4.0  
2               0.0  
3               1.0  
4               1.0  
5               0.0  
6               0.0  
7               1.0  
8               0.0  
9               0.0  
10              0.0  
11              0.0  
12              0.0  
13              1.0  
14              0.0  
15              0.0  
16              0.0  
17              1.0  
18              0.0  
19              0.0  
20              1.0  
21              0.0  
22              1.0  
23              1.0  
24              0.0  
25              0.0  
26              0.0  
27              3.0  
28              0.0  
29              0.0  
30              0.0  
31              1.0  
32              0.0  
33              1.0  
34              0.0  
35              0.0

#  Task 13 : 
# For the state of Maharashtra, what are the total number of seats, total number of candidates, total number of parties, total votes (including EVM and postal), and the breakdown of EVM and postal votes?


In [349]:
T13 = pd.read_sql_query("""select count(distinct cr.`Constituency ID`) as Total_Seats,
 count(distinct cd.`Candidate`) as Total_Candidates,
 count(distinct pr.`party`) as Total_Parties,
 sum(cd.`EVM Votes` + cd.`Postal Votes`) as  Total_Votes ,
 sum(cd.`EVM Votes`) as Total_EVM_Votes ,
 sum(cd.`Postal Votes`) as Total_Postal_Votes
 from constituencywise_results as cr
 join partywise_results as pr 
 on cr.`Party ID` = pr.`Party ID`
 join state_results as sr
 on cr.`Parliament Constituency`=sr.`Parliament Constituency`
 join states as s
 on s.`State ID` = sr.`State ID`
 join constituencywise_details as cd
 on cd.`Constituency ID` = cr.`Constituency ID` 
 where s.State='Maharashtra';
""", engine)

In [351]:
T13

Total_Seats  Total_Candidates  Total_Parties  Total_Votes  Total_EVM_Votes  \
0            3                76              2    3748130.0        3733135.0   

   Total_Postal_Votes  
0             14995.0

# task 14 : Which candidate received the highest number of EVM votes in each constituency (Top 10)?

In [372]:
T14 = pd.read_sql_query("""SELECT 
    cr.`Constituency Name`,
    cd.`Constituency ID`,
    cd.`Candidate`,
    cd.`EVM Votes`
FROM 
    constituencywise_details as  cd
JOIN 
    constituencywise_results as cr 
    ON cd.`Constituency ID` = cr.`Constituency ID`
WHERE 
    cd.`EVM Votes` = (
        SELECT MAX(cd1.`EVM Votes`)
        FROM constituencywise_details as cd1
        WHERE cd1.`Constituency ID` = cd.`Constituency ID`
    )
ORDER BY 
    cd.`EVM Votes` DESC 
    limit 10;

""", engine)

In [374]:
T14

Constituency Name Constituency ID                         Candidate  \
0    BANGALORENORTH           S1024                 SHOBHA KARANDLAJE   
1           RAIGARH            S262                RADHESHYAM RATHIYA   
2            RATLAM           S1224          ANITA NAGARSINGH CHOUHAN   
3        COOCHBEHAR            S251    JAGADISH CHANDRA BARMA BASUNIA   
4          RANAGHAT           S2513                  JAGANNATH SARKAR   
5        ANANTHAPUR           S0119  AMBICA G LAKSHMINARAYANA VALMIKI   
6          KOLHAPUR           S1347         CHHATRAPATI SHAHU SHAHAJI   
7             THANE           S1325              NARESH GANPAT MHASKE   
8          ULUBERIA           S2526                       SAJDA AHMED   
9          DINDIGUL           S2222                 SACHITHANANTHAM R   

   EVM Votes  
0     982805  
1     806667  
2     793432  
3     785747  
4     778046  
5     755862  
6     750323  
7     732109  
8     722586  
9     667391

# summary :

# 1.Parliament Constituency have total 543 Seats in INDIA GENERAL ELECTIONS RESULT ANALYSIS 2024

# 2.Total seats won by NDA alliance was 292 seats

# 3.NDA alliance parties like BJP won by 240